proposals/NNNN-generalize-async-sequence.md

```diff
@@ -0,0 +1,311 @@
+
+An expression is a non-optional derivation of an isolated parameter `param` if it is:
+
+* `param?` (the optional-chaining operator);
+* `param!` (the force-unwrapping operator); or
+* a reference to a _non-optional binding_ of `param`, i.e. a `let` constant initialized by a successful pattern-match which removes the optionality from `param`, such as `ref` in `if let ref = param`.
+
+When analyzing an argument expression in all cases above, certain non-instrumental differences in expression syntax and behavior must be ignored:
+
+* parentheses;
+* the effect-marking operators `try`, `try?`, `try!`, and `await`;
+* the type coercion operator as (in the cases where it doesn't perform a dynamic bridging conversion); and
+implicit type conversions such as promotion to `Optional` type.
+
+#### Calling `nextElement()` in async `for-in` loops
+
+Desugared async `for-in` loops will call `AsyncIteratorProtocol.nextElement()` instead of `next()` when the context has appropriate availability, and pass in an isolated argument value of type `(any Actor)?`. The argument value always represents the isolation of the caller so that the call does not cross an isolation boundary:
+
+* if the caller's context is `nonisolated`, the argument value is `nil`;
+* if the caller's context is isolated to some global actor `T`, the argument value is `T.shared`;
+* if the caller's context has an isolated parameter `param`, the argument value is `param`;
+* if the caller's context captures an isolated parameter `param`, the argument value is `param`.
+
+### Default implementations of `next()` and `nextElement()`
+
+Because existing `AsyncIteratorProtocol`-conforming types only implement `next()`, the standard library provides a default implementation of `nextElement()`:
+
+```swift
+extension AsyncIteratorProtocol {
+  /// Default implementation of `nextElement()` in terms of `next()`, which is
+  /// required to maintain backward compatibility with existing async iterators.
+  @available(SwiftStdlib 5.11, *)
+  @available(*, deprecated, message: "Provide an implementation of 'nextElement'")
+  public mutating func nextElement(_ actor: isolated (any Actor)?) async throws(Failure) -> Element? {
+    nonisolated(unsafe) var unsafeIterator = self
+    do {
+      let element = try await unsafeIterator.next()
+      self = unsafeIterator
+      return element
+    } catch {
+      throw error as! Failure
+    }
+  }
+}
+```
+
+Note that the default implementation of `nextElement` necessarily violates `Sendable` checking in order to pass `self` from a possibly-isolated context to a `nonisolated` one. Though this is generally unsafe, this is how calls to `next()` behave today, so existing conformances will maintain the behavior they already have. Implementing `nextElement()` directly will eliminate the unsafety.
```

Comment:
Of course, calls to `next()` today produce a warning under strict concurrency checking, would this default end up swallowing those warnings?

Reply:
Swift 5.9 doesn't diagnose calls to `next()`. The call wasn't diagnosed until https://github.com/apple/swift/pull/67730 which I then suppressed in https://github.com/apple/swift/pull/70750 because there's not a good workaround, and I've never seen anybody actually observe a data race on the iterator. In any case, the warnings are already swallowed so this proposal doesn't make any changes there.

Reply:
Ah great! Saw the warning when building against ~nightly at work but didn't realize it had been suppressed

Reply:
Ah great! Saw the warning when building against ~nightly at work but didn't realize it had been suppressed